### PR TITLE
Friendship -v2

### PR DESCRIPTION
**Pull request by gango-anan**
- This milestone involved implementing friendship using two for mutual friendship.

### CODE REVIEW

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:full_moon:
**Hi Galiwango!
Great job on this project:clap::clap:.**

**Major highlights**

- The tests widely cover both positive and negative scenarios
- There are no logic in view
- There is a button to accept and reject friendship in the users index page

**To improve**

- [ ] Bond model

- when friendship is confirmed, an additional row with reversed user-friend pair is created. Implement this by creating a method in the Bond model. eg 

`def  confirm_friend`
`self.update_attributes(confirmed: true)`
 `....`
`end`

- [ ] User Model

- User can find friends only by user_id in bond table. You already implemented this using ruby method. As shown below.

image

Now, the requirement for this milestone is to use SQL query to achieve the same. Why? since 
the database was created to store and query data it is always more efficient than ruby. The result should be similar to this:
```ruby
has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

- Also use SQL query for this pending-friend method.

image. Same reason as above.

- In order to get all posts written by user on their friends, use simple method in User model. Example

```ruby
def friends_and_own_posts
  Post.where(user: (self.friends.to_a << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

- [ ] Readme

- Add author name to the readme.

 [ Optional] 

- Commit more often as you make the above changes. So the PR can look more professional. Currently, we only have two commit messages for this PR.

Make the above-required changes and when you are ready, request another review.
Kindly drop your questions :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan.

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details close="close">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] Use better naming for relationships, something meaningful and descriptive. What does `indirect_friendship` tells you? And could it be better named and understand by anyone if you used something like `pending_friends`.
- [ ] Review previous Reviewer observations and change the methods for associations or scopes. 
Example: 
```
has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
``` 

</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [ ] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Comment by gango-anan** [comment]
Thanks for the reviews, requested changes made.

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -28,7 +28,9 @@ def notices_alerts
   end
 
   def invite_btn(user)
-    return if current_user.check_if_my_friend(user) || current_user == user
+    bonds = Bond.where(friend: current_user, user: user, confirmed: false)
+    friend = bonds.map(&:user)
+    return if current_user == user || friend.include?(user) || current_user.check_if_my_friend(user)
 
     button_to('Invite to Friendship', invite_path(friend_id: user.id), method: :post)
```
Comment (acushlakoncept) [inline app/helpers/application_helper.rb]: you should never use the `where` method, or any other query-building methods like this, outside of the model class itself, Take a look at this article for more Ruby on Rails best practices: [Buggy Rails Code: The 10 Most Common Mistakes That Rails Developers Make](https://www.toptal.com/ruby-on-rails/top-10-mistakes-that-rails-programmers-make)

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @gango-anan,  your level of progress on this project is indeed commendable, keep up the good work.

Before we proceed with any changes, keep this in mind. When deciding on what to name your methods, variables, class, or whatever, always think about other developers and your future self from 5-10 years from now. Will it be easy for them to understand what I meant with so and so name/method?

Also, try as much as possible to follow a given project naming convention. Coding is already a cognitively demanding task, let's not add to it by thinking what a method name means. 🙏 

- [ ] You did a great job with the `#timeline_posts` in the posts controller, but it could be better if we relieve our controller and let the model do the heavy lifting of handling most business logic. Calling this method from the model, instead of implementing the whole logic in the controller.
```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends.to_a << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```
Then in the `#timeline_posts`, we simple have `timeline_post.friends_and_own_posts`

- [ ] `User#pending_friends` in your case I believe `#unconfirmed_sent_requests` can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)
```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Bond", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```
In that case you will not need the defined ruby method any more `unconfirmed_sent_requests`

- [ ] This method 
```ruby
  def confirmed_friends
    direct_friendships.map { |friendship| friendship.friend if friendship.confirmed }.compact
  end
```
Since you already have `has_many :friends, through: :confirmed_friendships, source: :user` These are your confirmed friends, no need for the Ruby method.

- [ ] There should be NO application logic in the views nor helpers (no Post.new, Comment.create, current_user.friends.find_by(...)...). Everything connected with data processing should be encapsulated in models and used in controllers. I have made some comments on the affected method in your codebase below.

- [ ] Kindly delete the comment in your `posts_controller_spec.rb` file. It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
See this [post](https://stackoverflow.com/questions/13130859/inline-comments-in-ruby/13130900#:~:text=No%2C%20Ruby%20does%20not%20have,comment%20out%20the%20one%20row), about why comments are discouraged.

- [ ] Your ReadMe file still carries the app template description. "**This repo includes initial code for social media app with basic styling. Its purpose is to be a starting point for Microverse students.**" I am sure you will agree with me, this repo no longer carries initial code, you sure have done a lot. Do well to update it to reflect what the app is all about. Possibly, a screenshot too, to give it that professional look.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by NtwaliHeritier — APPROVED** [review]
### **STATUS: APPROVED✔✔✔**

Hello @gango-anan ,
Great work on this milestone.
You can merge and move on👍👍

Happy coding💻💻💻

**Pull request merged**